lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp

```diff
@@ -2963,6 +2963,252 @@ class DecomposeAtenLinearOp : public OpRewritePattern<AtenLinearOp> {
+              dotResultShape.push_back(operandShape[en.index()]);
+            }
+          }
+        };
+
+    for (int i = 0; i < totalRank - 1; i++) {
+      SmallVector<int64_t> lhsContractingDims, lhsBatchingDims,
+          rhsContractingDims, rhsBatchingDims;
+      SmallVector<char> dotResultTokens;
+      SmallVector<int64_t> dotResultShape;
+      SmallVector<int64_t> lhsShape = inputShapes[0];
+      SmallVector<int64_t> rhsShape = inputShapes[1];
+      SmallVector<char> lhsTokens, rhsTokens;
+      lhsTokens = inputTokens[0];
+      rhsTokens = inputTokens[1];
+      collectOperandDims(lhsShape, lhsTokens, rhsTokens, lhsContractingDims,
+                        lhsBatchingDims, dotResultTokens, dotResultShape);
+      collectOperandDims(rhsShape, rhsTokens, lhsTokens, rhsContractingDims,
+                        rhsBatchingDims, dotResultTokens, dotResultShape);
+      // Prepend batch tokens.
+      for (const auto &it : llvm::enumerate(lhsBatchingDims)) {
+        char batchingToken = lhsTokens[it.value()];
+        int64_t batchingShapeDim = lhsShape[it.value()];
+        dotResultTokens.insert(dotResultTokens.begin() + it.index(),
+                              batchingToken);
+        dotResultShape.insert(dotResultShape.begin() + it.index(),
+                              batchingShapeDim);
+      }
+      // Lowering to dot_general does not support a mismatch between the number
+      // of result dims and the number of non-contracting dims.
+
+      if (!lhsContractingDims.empty()) {
+        int rank = lhsTokens.size() - 1;
+        if (lhsContractingDims[0] != rank) {
+          Value zero =
+              rewriter.create<ConstantIntOp>(loc, rewriter.getI64IntegerAttr(lhsContractingDims[0]));
+          Value one =
+              rewriter.create<ConstantIntOp>(loc, rewriter.getI64IntegerAttr(rank));
+          size_t temp = lhsShape[lhsContractingDims[0]];
+          int rank = lhsTokens.size() - 1;
+          lhsShape[lhsContractingDims[0]] = lhsShape[rank];
+          lhsShape[lhsTokens.size() - 1] = temp;
+          torchTensors[0] = rewriter.create<AtenTransposeIntOp>(loc, op.getType(),torchTensors[0], zero, one);
+        }
+      } else {
+        lhsShape.insert(lhsShape.begin() + lhsTokens.size(), 1);
+      }
+      if (!lhsBatchingDims.empty()) {
+        if (lhsBatchingDims[0] != 0) {
+          Value zero =
+              rewriter.create<ConstantIntOp>(loc, rewriter.getI64IntegerAttr(lhsBatchingDims[0]));
+          Value one =
+              rewriter.create<ConstantIntOp>(loc, rewriter.getI64IntegerAttr(0));
+          torchTensors[0] = rewriter.create<AtenTransposeIntOp>(loc, op.getType(),torchTensors[0], zero, one);
+          size_t temp = lhsShape[lhsBatchingDims[0]];
+          lhsShape[lhsBatchingDims[0]] = lhsShape[0];
+          lhsShape[0] = temp;
+        }
+      } else {
+        lhsShape.insert(lhsShape.begin(), 1);
+      }
+
+      if (!rhsContractingDims.empty()) {
```

Comment:
Will take a closer look tomorrow, but this pattern has multiple cases, which should be accompanied by lit tests making sure each case works as expected: https://github.com/llvm/torch-mlir/blob/main/docs/architecture.md#dos-and-donts-for-unit-vs-end-to-end-testing

Reply:
I accidentally closed this PR, I will reopen one and add more test cases to make sure every case is tested